### DISCUSSION TITLE
Custom Line Items Without Storing Products in MedusaJS Backend

### DISCUSSION THREAD

**Discussion by jkuzmanovik** [medusajs/medusa#9538] [Q&A]
Problem:
I'm building an eCommerce site using MedusaJS v2 for the backend and a frontend framework (like Next.js). The products I'm working with are fetched from an external API and not stored in my Medusa backend. The reason is that these products change frequently in availability, pricing, and other details, and there are millions of them, which makes it impractical to store them directly in the Medusa product catalog.

However, I still want to allow users to add these external products to the cart and proceed to checkout through MedusaJS. But as I understand, Medusa requires products and variants to be stored in the backend in order to add them to the cart.

What I've Tried:
Fetching product data dynamically from the external API: This works well for displaying products on the frontend, but when trying to add them to the cart, I run into issues since they aren't stored in the Medusa backend.
Adding custom line items: I came across a note in the documentation that says a custom line item can be added to the cart without a variant. However, I couldn't find any detailed instructions or examples on how to implement this, and attempts to add custom items without a variant_id result in errors.
The Solution I'm Looking For:
I need a way to add dynamically fetched products to the cart without storing them as products or variants in the Medusa backend. Ideally, I want to be able to:

Add a custom item to the cart by passing product information (e.g., price, description, metadata) directly to the POST /store/carts/:id/line-items API.
Skip creating or storing the product/variant in the backend for each item, as the data will be fetched in real-time from the external source.
Questions:
Is it possible to bypass the need for storing products in the Medusa backend and use custom line items to handle this kind of use case?
If the API enforces a variant_id, is there a workaround to treat a placeholder variant as a "shell" for custom items and store product data (such as price, description, etc.) in the metadata field? Is this an intended use case for Medusa?
Are there any plans to introduce better support for scenarios where products are fetched externally but still need to go through the cart and checkout flow?
Any guidance on how I can achieve this would be greatly appreciated!

**Comment by radoslaw-sz** [comment]
hi @jkuzmanovik  - I do not know if it is yet available, but Medusa v2 contains a lot of commerce modules, including the Product Module: https://docs.medusajs.com/v2/resources/commerce-modules/product. Probably you would need to completely replace this package by your own module, so it means you would need to implement all logic related to product module, including methods which are needed there.
I think it is the only option here.

**Comment by olivermrbl** [comment]
@jkuzmanovik, appreciate the interesting in Medusa.

I've been working on enabling custom line items for carts and orders more natively by decoupling them entirely from the product variants. 

The work-in-progress PR is [here](https://github.com/medusajs/medusa/pull/10408).

The changes in the PR should enable the use case you are describing, but I'd love it if you could test the snapshot version in the PR. I tried to replicate your flow locally with the following code–perhaps this can serve as inspiration:
```ts
// src/api/store/carts/[id]/custom-line-items/route.ts

import { addToCartWorkflow } from "@medusajs/medusa/core-flows";

// You need to replace this with a call to your own third-party products API
const thirdPartyCall = async (item) => {
  return {
    title: "Some other product",
    sku: "1234567890",
    variant_barcode: "1234567890",
    unit_price: 100,
  };
};

export const POST = async (req, res) => {
  const { id } = req.params;
  const { items } = req.body;

  const query = req.scope.resolve("query");

  const { data } = await query.graph({
    entity: "cart",
    filters: { id },
    fields: ["id", "sales_channel_id"],
  });

  const itemsWithExternalProducts = await Promise.all(
    items.map(async (item) => {
      const someExternalProductVariant = await thirdPartyCall(item);
      return {
        ...item,
        title: someExternalProductVariant.title,
        variant_sku: someExternalProductVariant.sku,
        variant_barcode: someExternalProductVariant.variant_barcode,
        unit_price: someExternalProductVariant.unit_price,
      };
    })
  );

  const workflowInput = {
    items: itemsWithExternalProducts,
    cart: data[0],
  };

  await addToCartWorkflow(req.scope).run({
    input: workflowInput,
  });

  const updatedCart = await query.graph({
    entity: "cart",
    filters: { id },
  });

  res.status(200).json({ cart: updatedCart });
};

```

The main idea here is to re-use the `addToCartWorkflow` from our core and pass custom line item data that is generated from your request to your external products API. Please note, you should not allow prices to be passed directly to the endpoint, as this would effectively mean that customers can price their own products. Instead this should also be computed/retrieved on the server.


Let me know how it goes.
